community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/internal/ResetFuzzTest.java

```diff
@@ -0,0 +1,282 @@
+    // ahead in the message queue, we use this test to convince ourselves
+    // there is no code path where RESET causes a session to not go back
+    // to a good state.
+
+    private final int seed = new Random().nextInt();
+
+    private final Random rand = new Random( seed );
+    private final LifeSupport life = new LifeSupport();
+    /** We track the number of un-closed transactions, and fail if we ever leak one */
+    private final AtomicLong liveTransactions = new AtomicLong();
+    private final Neo4jJobScheduler scheduler = life.add(new Neo4jJobScheduler());
+    private final SessionStateMachine ssm = new SessionStateMachine( new FuzzStubSPI() );
+    private final ThreadedSessions sessions =
+            new ThreadedSessions( ( enc ) -> ssm, scheduler, NullLogService.getInstance() );
+
+    private final List<Message> messages = asList(
+        new RunMessage( "test", map() ),
+        new DiscardAllMessage(),
+        new PullAllMessage(),
+        new ResetMessage()
+    );
+
+    private final List<Message> sent = new LinkedList<>();
+
+    @Test
+    public void shouldAlwaysReturnToIdleAfterReset() throws Throwable
+    {
+        // given
+        life.start();
+        Session session = sessions.newSession();
+        session.init( "Test/0.0.0", map(), null, Session.Callback.NO_OP );
+
+        TransportBridge bridge = new TransportBridge(
+                NullLog.getInstance(), session, new MessageHandler.Adapter<>(), (() -> {} ) );
+
+        // 5 seconds lead to testing ~1M permutations on my 5-year old MBP, so
+        // 2 seconds seemed like a sensible balance, testing 300K permutations or
+        // so per run without taking up too much test time. Simply bump this to
+        // run a longer test. This was green at 300 second runs when this is written.
+        long deadline = System.currentTimeMillis() + 2 * 1000;
+
+        // when
+        while( System.currentTimeMillis() < deadline )
+        {
+            dispatchRandomMessages( bridge );
+            assertSessionWorks( session );
+        }
+    }
+
+    private void assertSessionWorks( Session session )
+    {
+        RecordingCallback recorder = new RecordingCallback();
+        session.reset( null, recorder );
+
+        try
+        {
+            assertThat( recorder, recorded( success() ) );
+            assertThat( ssm.state(), equalTo( IDLE ) );
+            assertThat( liveTransactions.get(), equalTo( 0L ));
+        }
+        catch( AssertionError e )
+        {
+            throw new AssertionError( String.format("Expected session to return to good state after RESET, but " +
+                                                    "assertion failed: %s.%n" +
+                                                    "Seed: %s%n" +
+                                                    "Messages sent:%n" +
+                                                    "%s",
+                    e.getMessage(), seed, Iterables.toString(sent, "\n") ), e );
```

Comment:
Code style: should there be a space after opening parens? Same situation in line 99 and 128.